Create four base grid patterns

mrdavenguyen opened the pull request:
- Modify grid generation algorithm to fill every line instead of alternating lines
	- Modify find_usable_space to only iterate through the first half of the middle lines in order to maintain symmetry
- Create function to generate four different offsets of the base grid pattern
- Update CHANGELOG with feature

mrdavenguyen committed:
Create four base grid patterns to allow for more variation in grid designs

- Modify grid generation algorithm to fill every line instead of alternating lines
	- Modify find_usable_space to only iterate through the first half of the middle lines in order to maintain symmetry
- Create function to generate four different offsets of the base grid pattern
- Update CHANGELOG with feature

mrdavenguyen committed:
Remove unnecessary parameters from find_usable_space

- Place variables inside function instead

mrdavenguyen committed:
Update find_usable_space with dataclass

- Replace variables with dataclass in:
	- find_usable_spaces
	- find_first_space
	- find_last_space_rows
	- find_last_space_cols

mrdavenguyen committed:
Merge branch 'main' into 23-grid-class-create-4-base-grid-patterns-to-allow-for-more-variation-in-grid-designs

crissycancode commented on grid.py:
```diff
@@ -1,13 +1,17 @@
-from typing import List
-from typing import Optional
-from typing import Tuple
+from typing import Callable, Dict, List, Optional, Tuple
```
I'd have to agree on this. Looks good!

crissycancode commented on grid.py:
```diff
@@ -428,27 +450,24 @@ def populate_lines(self, orientation: str) -> None:
-    def create_alternating_divisions(self, row: int) -> None:
+                    self.place_center_divider_randomly(half_grid)
+                
+    def place_center_divider_randomly(self, half_grid: int) -> None:
```
Can we change the name to something closer to the description? The function name is quite vague but the description is on point. Maybe words from the description can be borrowed? Suggestion for clarity

crissycancode commented on grid.py:
```diff
@@ -557,7 +576,7 @@ def create_random_word_lengths(self, remaining_space: int, num_words: int, word_
             word_lengths.append(word_length)
     
-    def find_usable_spaces(self, line: int, orientation: str, is_space: bool = False, first_space: int = 0, last_space: int = 0) -> List[List[int]]:
+    def find_usable_spaces(self, line: int, orientation: str, half_grid: int) -> List[List[int]]:
```
❤️

crissycancode commented on grid.py:
```diff
@@ -611,19 +631,18 @@ def find_last_space_rows(self, pos: int, is_space: bool, first_space: int, last_
+                    usable_spaces.append((self.usable_space.first_space, self.usable_space.last_space))
     
-    def find_last_space_cols(self, pos: int, is_space: bool, first_space: int, last_space: int, line: int, usable_spaces: List[List[int]]) -> bool:
+    def find_last_space_cols(self, pos: int, line: int, usable_spaces: List[List[int]]) -> None:
```
If the function is being called per column, maybe col instead of cols. Like how we only use plural form for lists, arrays, dict, any collection type

crissycancode reviewed (approved):
👏

mrdavenguyen committed:
Implement changes based on feedback

- Change function name from place_center_divider_randomly to conditionally_place_center_divider
